### DISCUSSION TITLE
Please provide guidance with Docker usage

### DISCUSSION THREAD

**Discussion by opyate** [prisma/prisma#23696] [Q&A]
### Question

Just seeing [this error](https://github.com/prisma/prisma/issues/20479) for the first time, and it raises the question - why write to `app/node_modules/*` in the first place?

When a Docker image is created, usually you run `npm i` as root then later on designate a non-privileged user with which to run the app. I think the disconnect is here: `root` owns everything in `node_modules`, but my `node` user doesn't have permissions to write to anywhere in the `node_modules` tree.

I tried searching the prisma website for some guidance on using with Docker, but couldn't find any, if you could kindly point me in the right direction.

E.g. I tried adding `RUN npm install -g prisma` before the line `USER node` and removed prisma dependency from my package.json, but the `Can't write to ???/node_modules/prisma` issue persists.

**Comment by ludralph** [comment]
Hi @opyate 

The error you're encountering is related to the permissions of the `node_modules` directory in your Docker container. When you install packages as root, they are owned by the root user. If you later switch to a non-root user (like node), this user may not have the necessary permissions to write to the `node_modules` directory.

Prisma needs to write to the `node_modules` directory because it generates a Prisma Client, which is a type-safe database client that's tailored to your database schema. This client is generated in your `node_modules` at `@prisma/client`.

To resolve this issue, you can try changing the ownership of the `node_modules` directory to the node user before running the Prisma commands. You can do this by adding the following line to your Dockerfile:
```
RUN chown -R node:node /path/to/node_modules
```
Replace `/path/to/node_modules` with the actual path to your `node_modules` directory.

In addition, you might want to specify a [custom `output`](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/generating-prisma-client#using-a-custom-output-path) directory for the Prisma Client if you can't modify the permissions of the `node_modules` directory. You can do this in your Prisma schema file as follows:
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "../custom/path"
}
```
Replace `../custom/path` with your desired output path. 

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
